notebooks/03-query/15-pivot-unpivot.ipynb

```diff
@@ -0,0 +1,372 @@
+    "\n",
+    "\n",
+    "The following query uses the results of the pivot to then aggregate totals for the pivot columns by channel. The aggregations being calculated could be NULL in the case where \"isRobot\"='true' (or 'false') does not exist for a given channel. The query calculates a ratio of human to total changes for each channel and the result is sorted on this ratio such that the channels with the highest proportion of robot updates are listed first. So in order to calculate this metric even in presence of NULLs the query uses COALESCE on all the pivoted metrics.\n",
+    "\n",
+    "The result is much cleaner than the prior query and it is easy to determine which channels have the most robotic activity."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65c489a8-a7f2-4cc0-b30c-f30d8ed32be2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT \"channel\", \n",
+    "        SUM(\"robot_added\")       AS \"added_robot\", \n",
+    "        SUM(\"robot_deleted\")     AS \"deleted_robot\", \n",
+    "        SUM(\"human_added\")       AS \"added_human\",\n",
+    "        SUM(\"human_deleted\")     AS \"deleted_human\", \n",
+    "        \n",
+    "        SAFE_DIVIDE( SUM(\"human_deleted\" + \"human_added\") * 1.0 , \n",
+    "                     SUM( \"robot_deleted\" + \"robot_added\" + \"human_deleted\" + \"human_added\" )\n",
+    "                     \n",
+    "                    )            AS \"human_ratio\"\n",
+    "                    \n",
+    "FROM\n",
+    "(\n",
+    "    SELECT\n",
+    "      \"channel\", \n",
+    "       COALESCE(\"robot_added\",0) AS \"robot_added\", \n",
+    "       COALESCE(\"human_added\",0) AS \"human_added\", \n",
+    "       COALESCE(\"robot_deleted\",0) AS \"robot_deleted\", \n",
+    "       COALESCE(\"human_deleted\",0) AS \"human_deleted\"\n",
+    "    FROM \"example-pivot-unpivot\" \n",
+    "    \n",
+    "    PIVOT ( SUM(added) as added, \n",
+    "            SUM(deleted) as deleted \n",
+    "            \n",
+    "            FOR \"isRobot\" IN ('true' as robot, 'false' as human)\n",
+    "          )\n",
+    ")x\n",
+    "GROUP BY 1\n",
+    "ORDER BY 6\n",
+    "'''\n",
+    "display_client.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3f0af833-9508-4f26-a484-2778ca857a2a",
+   "metadata": {},
+   "source": [
+    "## UNPIVOT example\n",
+    "Given the prior query we can now investigate the sources of the updates in the most robotically updated pages.\n",
+    "```\n",
+    "  UNPIVOT ( \"changes\"  FOR \"action\" IN (\"added\", \"deleted\") )\n",
+    "```\n",
+    "This example uses UNPIVOT to:\n",
+    "- take the values of multiple columns: `\"added\",\"deleted\"`\n",
+    "- incorporate them into a single column called the \"values column\": `\"changes\"`\n",
+    "- while expanding the result with a \"names column\" containing the names of the pivoted columns: `\"action\"`\n",
```

Comment:
nit "while" so it's just "expand the result with.."

Reply:
done